### PR TITLE
Vaccination report7

### PR DESCRIPTION
**Pull request by JennaVergeynst**
@mrollier and @twallema please check the description: some important changes in this PR!

**Novelties** (both to the non-spatial and spatial model):
- Injection of the new strain is now done inside the model, by providing parameters injection_day and injection_ratio. This avoids running the model until injection day and taking the average of the states at that point (because at that point you lose uncertainty).
- dalpha is protected against becoming nan (when division error occurs)
- new state VE = sum of all vaccination eligible states => needed in the time-dependent vaccination function
- vaccine leakiness added: when parameter leakiness is provided, a part of the vaccinated people still contributes to infecting susceptibles.
- time-dependent functions now always have to start with def fun(t, states, param, …), otherwise an error message appears. So now the states of the model are available in the time-dependent functions (was needed for the vaccination function).

**To do:**
There is still a problem between the spatial model and vaccinations, either in the implementation or in the vaccination function. This should not affect any model run without vaccination though.
I will fix this after my break.

### CODE REVIEW

File: src/covid19model/models/models.py [hunk 1]
```diff
@@ -149,36 +152,45 @@ def integrate(t, S, E, I, A, M, ER, C, C_icurec, ICU, R, D, H_in, H_out, H_tot,
         # calculate total population
         # ~~~~~~~~~~~~~~~~~~~~~~~~~~
         T = S + E + I + A + M + ER + C + C_icurec + ICU + R + V
-        vacc_eligible = S + R + E + I + A
+        # vaccination eligible states
+        VE = S + R + E + I + A
```
Comment (twallema) [inline src/covid19model/models/models.py:156]: Is it necessary to compute VE here? In terms of bookkeeping, VE is a state which is updated using the differential dVE, so this doesn't seem necessary.

File: src/covid19model/models/models.py [hunk 2]
```diff
@@ -149,36 +152,45 @@ def integrate(t, S, E, I, A, M, ER, C, C_icurec, ICU, R, D, H_in, H_out, H_tot,
...
-        dV_new = N_vacc/vacc_eligible*S + N_vacc/vacc_eligible*R + N_vacc/vacc_eligible*E + N_vacc/vacc_eligible*I + N_vacc/vacc_eligible*A - V_new
-        dV = N_vacc/vacc_eligible*S + N_vacc/vacc_eligible*R + N_vacc/vacc_eligible*E + N_vacc/vacc_eligible*I + N_vacc/vacc_eligible*A - (IP_old + IP_new)*(1-e)*V
-        # If A and I are both zero, a division error occurs
+        dV_new = N_vacc/VE*S + N_vacc/VE*R + N_vacc/VE*E + N_vacc/VE*I + N_vacc/VE*A - V_new
+        dV = N_vacc/VE*S + N_vacc/VE*R + N_vacc/VE*E + N_vacc/VE*I + N_vacc/VE*A - (IP_old + IP_new)*(1-e)*V
+        dVE = dS + dR + dE + dI + dA
         # Update fraction of new COVID-19 variant
         dalpha = IP_new/(IP_old+IP_new) - alpha
+         # If A and I are both zero, a division error occurs
+        dalpha[np.isnan(dalpha)] = 0
```
Comment (twallema) [inline src/covid19model/models/models.py:187]: When using one exposed individual as the initial condition, both `IP_old` and `IP_new` are zero, doesn't the algorithm then 'get stuck' while trying to compute dalpha? I.e. does this protection work in practice? If not, maybe a `try`, `except` structure can help?

**Review by twallema — APPROVED** [review]
Mooie implementatie! Ik heb twee vraagjes, zie comments.

**Comment by JennaVergeynst** [comment]
Damn, ik zie nu pas je comments, heb al gemerged

File: src/covid19model/models/models.py (hunk 1 above)
Comment (JennaVergeynst) [reply src/covid19model/models/models.py:156]: true

File: src/covid19model/models/models.py (hunk 2 above)
Comment (JennaVergeynst) [reply src/covid19model/models/models.py:187]: mmh... good question... something to keep in mind

**Comment by mrollier** [comment]
Very clear implementation, I think. Some general remarks and questions below. Please read through them and verify that I understand what is going on.

- injection_day injection_ratio are both 0 by default in get_COVID_SEIRD_parameters
- There is a degeneracy between injection_day and injection_ratio, I think: early injection with low injection ratio should be the same (right?) as late injection with high injection ratio
- Not sure what the original problem was "running the model until injection day and taking the average of the states at that point"
- Infectivity gain is set to K=0 by default. Shouldn't this be K=1?
- So alpha must be initialised as an age- and spatially stratified matrix with floating points between 0 and 1.
	If it is zero, it stays zero, unless if we pass injection_day and inject injection_ratio (which is also a value between 0 and 1).
	Note: this doesn't happen if alpha has been initialised, so there's a choice to be made: either initialise alpha, OR initialise it dynamically by
	defining injection_day and injection_ratio. The second approach allows for more control, but is not entirely realistic (there's not "suddenly" 10 percent VOC ratio)
- No implementation for stochastic case yet (not urgent at all)
- How can this be generalised to an arbitrary number of VOCs?
- It is currently assumed that vaccination works equally well for "classic" SARS-CoV-2 vs. all VOCs
- In the spatial model, dalpha is written more concisely: "dalpha = alpha*K/(1-alpha+alpha*K) - alpha". This can also be above, I suppose
- Leakiness is supposed 0 for all ages. This is not yet based on data, I suppose?
- "states" and "param" are not explained in the description of the functions in time_dependant_parameter_fncs.py

**Comment by twallema** [comment]
> Very clear implementation, I think. Some general remarks and questions below. Please read through them and verify that I understand what is going on.
> 
> * injection_day injection_ratio are both 0 by default in get_COVID_SEIRD_parameters
> * There is a degeneracy between injection_day and injection_ratio, I think: early injection with low injection ratio should be the same (right?) as late injection with high injection ratio
> * Not sure what the original problem was "running the model until injection day and taking the average of the states at that point"
> * Infectivity gain is set to K=0 by default. Shouldn't this be K=1?
> * So alpha must be initialised as an age- and spatially stratified matrix with floating points between 0 and 1.
>   If it is zero, it stays zero, unless if we pass injection_day and inject injection_ratio (which is also a value between 0 and 1).
>   Note: this doesn't happen if alpha has been initialised, so there's a choice to be made: either initialise alpha, OR initialise it dynamically by
>   defining injection_day and injection_ratio. The second approach allows for more control, but is not entirely realistic (there's not "suddenly" 10 percent VOC ratio)
> * No implementation for stochastic case yet (not urgent at all)
> * How can this be generalised to an arbitrary number of VOCs?
> * It is currently assumed that vaccination works equally well for "classic" SARS-CoV-2 vs. all VOCs
> * In the spatial model, dalpha is written more concisely: "dalpha = alpha_K/(1-alpha+alpha_K) - alpha". This can also be above, I suppose
> * Leakiness is supposed 0 for all ages. This is not yet based on data, I suppose?
> * "states" and "param" are not explained in the description of the functions in time_dependant_parameter_fncs.py

- Yes, this results in a model without vaccination.
- No, even with very low injection prelevance (10^-6), the British variant overtakes rapidly. I.e. the sensitivity to the injection day is much higher than to the infection prelevance.
-  Run model until introduction of british variant --> take states at last timestep --> compute average --> re-initialize model with initial condition equal to last timestep --> initialize british variant --> run simulation untill desired enddate. This way, all uncertainty at the last timestep before introduction of the new strain is reduced to a single point, the mean.
- Infectivity gain should indeed be set to zero by default, makes more sense.
- `alpha` is a model state, and is thus, for the national model, age-stratified, and for the spatial model age- and space-stratified.
- Agree with the following two bullet points.
- With regard to a generalization with an arbitrary number of VOCs, this is a very good remark. An option would be to add yet another 'axis' to the model, the VOC axis, but this will probably slow down calculations too much. As an alternative, the equation for the 'infection pressure' (`IP`) could easily be n-dimensionalized by adding an extra VOC axis to the state `alpha` and to the gain `K`, and then computing an n-dimensional weighted average for `IP`. In this way, not every state needs an extra VOC axis. The question is how we could manage this in `base.py` in a manner that is sufficiently elegant and general.

**Comment by JennaVergeynst** [comment]
- leakiness parameter: this enables to add leakiness to the scenario's, but as long as it is not used in scenario's, the parameter is 0.
- states and params: right, I'll add this!

**Pull request merged**